File: src/lib/tool-manifest/tools/get-image-context.ts
import { z } from 'zod';
import { useAiSession } from '@/hooks/useImageContext';
import type { ToolManifest } from '../types';

const input = z.object({}).describe('No input — returns the cached ImageContext.');

const output = z.object({
  available: z.boolean(),
  context: z.unknown().nullable().describe('The ImageContext object, or null if no image has been analysed yet.'),
});

export const getImageContextTool: ToolManifest<typeof input, typeof output> = {
  name: 'get_image_context',
  kind: 'query',
  description:
    'Read the cached image analysis (subjects, lighting, mood, dominant tones, candidate regions). Use this first to understand what kind of image you are editing.',
  inputSchema: input,
  outputSchema: output,
  handler: () => {
    const ctx = useAiSession.getState().context;
    return { available: ctx !== null, context: ctx ?? null };
  },
};

File: backend/app/tools/atomic/get_image_context.py
from __future__ import annotations

from pydantic import BaseModel

from app.state.document import SessionDocument
from app.tools.base import BackendTool, ToolPermissions


class _Input(BaseModel):
    pass


class _Output(BaseModel):
    available: bool
    context: dict | None


class GetImageContextTool(BackendTool[_Input, _Output]):
    name = "get_image_context"
    kind = "query"
    description = (
        "Read the cached image analysis (subjects, lighting, mood, dominant tones, "
        "candidate regions). Call this first to understand the photo."
    )
    input_schema = _Input
    output_schema = _Output
    permissions = ToolPermissions(requires_image=True, requires_context=False)

    async def handler(self, doc: SessionDocument, input: _Input) -> _Output:  # noqa: A002
        ctx = doc.image_context
        if ctx is None:
            return _Output(available=False, context=None)
        return _Output(available=True, context=ctx.model_dump(mode="json"))
